simpleGraph/search.go

```diff
@@ -1,107 +1,637 @@
+			}
+			if readBase == dnaTwoBit.GetBase(node.Prev[i].Dest.SeqTwoBit, uint(node.Prev[i].Dest.SeqTwoBit.Len)-1) {
+				prevParts = extendToTheLeftHelperDev(node.Prev[i].Dest, read, currPart)
+				answer = append(answer, prevParts...)
+			}
+		}
+	}
+
+	if len(answer) == 0 {
+		return []SeedDev{currPart}
+	} else {
 		return answer
+	}
+}
+
+func extendToTheLeftHelperDev(node *Node, read *fastq.FastqBig, nextPart SeedDev) []SeedDev {
+	const basesPerInt int = 32
+	var nodePos int = node.SeqTwoBit.Len - 1
+	var readPos int = int(nextPart.QueryStart) - 1
+	var nodeOffset int = nodePos % basesPerInt
+	var readOffset int = 31 - ((readPos - nodeOffset + 31) % 32)
+	var leftMatches int = 0
+	var currPart SeedDev
+	var prevParts, answer []SeedDev
+	var i int
+	var readBase dna.Base
+
+	if nextPart.PosStrand {
+		leftMatches = common.Min(readPos+1, dnaTwoBit.CountLeftMatches(node.SeqTwoBit, nodePos, &read.Rainbow[readOffset], readPos+readOffset))
+	} else {
+		leftMatches = common.Min(readPos+1, dnaTwoBit.CountLeftMatches(node.SeqTwoBit, nodePos, &read.RainbowRc[readOffset], readPos+readOffset))
+	}
+
+	if leftMatches == 0 {
+		log.Fatal("Error: should not have zero matches to the left\n")
+	}
+	currPart = SeedDev{TargetId: node.Id, TargetStart: uint32(nodePos - (leftMatches - 1)), QueryStart: uint32(readPos - (leftMatches - 1)), Length: uint32(leftMatches), PosStrand: nextPart.PosStrand, TotalLength: uint32(leftMatches) + nextPart.TotalLength, NextPart: &nextPart, Next: nil}
+	// we went all the way to end and there might be more
+	if currPart.QueryStart > 0 && currPart.TargetStart == 0 {
+		for i = 0; i < len(node.Prev); i++ {
+			if nextPart.PosStrand {
+				readBase = dnaTwoBit.GetBase(&read.Rainbow[0], uint(currPart.QueryStart)-1)
+			} else {
+				readBase = dnaTwoBit.GetBase(&read.RainbowRc[0], uint(currPart.QueryStart)-1)
+			}
+			if readBase == dnaTwoBit.GetBase(node.Prev[i].Dest.SeqTwoBit, uint(node.Prev[i].Dest.SeqTwoBit.Len)-1) {
+				prevParts = extendToTheLeftHelperDev(node.Prev[i].Dest, read, currPart)
+				answer = append(answer, prevParts...)
+			}
+		}
+	}
+	// if the alignment did not go to another node, return the match for this node
+	if len(answer) == 0 {
+		answer = []SeedDev{currPart}
+	}
+	return answer
+}
+
+func newSeedBuilder() *seedHelper {
+	var tmp SeedDev = SeedDev{}
+	return &seedHelper{
+		currHits: make([]uint64, 0, 20),
+		tempSeed: tmp,
+	}
+}
+
+func restartSeedHelper(helper *seedHelper) {
+	helper.currHits = helper.currHits[:0]
+	helper.keyIdx, helper.keyOffset, helper.readOffset, helper.nodeOffset = 0, 0, 0, 0
+	helper.nodeIdx, helper.nodePos = 0, 0
+	helper.seqKey, helper.codedNodeCoord = 0, 0
+	helper.leftMatches = 0
+}
+
+//seedBuildHelper.nodeIdx, seedBuildHelper.nodePos int64 = 0, 0
+func seedMapMemPool(seedHash map[uint64][]uint64, nodes []*Node, read *fastq.FastqBig, seedLen int, perfectScore int64, scoreMatrix [][]int64, finalSeeds []SeedDev, tempSeeds []SeedDev, seedBuildHelper *seedHelper) []SeedDev {
```

Comment:
Are there changes beyond reusing memory.  It takes me a little time to reason through all the offsets.  I am not sure if there are changes there or not